### DISCUSSION TITLE
Custom comparison

### DISCUSSION THREAD

**Discussion by joshua-light** [dotnet/fsharp#13177] [Q&A] [answered]
Hi!
I'm trying to design a safe type called `Nat` that represents a natural number (1, 2, 3, etc.). Something like this:
```fs
type Nat =
    private | Nat of int
    
    static member One = Nat 1
    
    static member New (x: int) : Nat option =
        if x > 0
        then Some (Nat x)
        else None
        
    static member (+) (Nat a, Nat b) = Nat (a + b)
    static member (-) (Nat a, Nat b) = Nat.New (a - b)
    static member (*) (Nat a, Nat b) = Nat (a * b)
    static member (/) (Nat a, Nat b) = Nat (a / b)
```

Since this type is a simple `int` wrapper, it's reasonable to define some operators overloaded for `int`, for example:
```fs
static member (+) (Nat a, b: int)  = Nat.New (a + b)
```

Now consider you have an object `x` with field `Count: Nat` and you simply want to check whether `x.Count > 1`. It seems like it's impossible (but makes sense) to overload equality or comparison operator:
```fs
static member (=) (Nat a, b: int) = a = b // Error
```

Is there any workaround to do something like that?
This can be solved partially by implementing implicit cast from `Nat` to `int` (which is totally reasonable) and then comparing in reverse order (e.g. `1 < x.Count`). But it's not readable in the same way as the first option.

Thanks in advance!

_P.S. While it's perfectly possible to define a custom operator or a function (e.g. `x.Count |> gt 1`), it'd be much more convenient to use the same syntax for custom types without additional noise._

**Comment by dsyme** [comment, marked as answer]
There is no direct way to overload`=` in F#.  You could define your own operator.  You might also get some mileage from an `op_Implicit`.  However neither are really that recommended (they make code harder to understand, and you should probably just really accept a tax of an explicit coercion for converting `int` --> `Nat`.
